src/menu.cc

```diff
@@ -105,106 +105,181 @@ extern "C" int nm_menu_hook(void *libnickel, nm_menu_item_t **items, size_t item
     QString trrm = QCoreApplication::translate("DictionaryActionProxy", "Dictionary");
     NM_LOG("Comparing against '%s', '%s'", qPrintable(trmm), qPrintable(trrm));
 
-    bool ismm, isrm; // only one will be true
-    if ((ismm = (label == trmm) && !checkable))
+    nm_menu_location_t loc = {};
+    if (label == trmm && !checkable) {
         NM_LOG("Intercepting main menu (label=Settings, checkable=false)...");
-    if ((isrm = (label == trrm) && !checkable))
+        loc = NM_MENU_LOCATION_MAIN_MENU;
+    } else if (label == trrm && !checkable) {
         NM_LOG("Intercepting reader menu (label=Dictionary, checkable=false)...");
+        loc = NM_MENU_LOCATION_READER_MENU;
+    }
+
+    if (loc)
+        QObject::connect(menu, &QMenu::aboutToShow, std::bind(_nm_menu_inject, _this, menu, loc, menu->actions().count()));
+
+    return AbstractNickelMenuController_createMenuTextItem_orig(_this, menu, label, checkable, checked, thingy);
+}
+
+void _nm_menu_inject(void *nmc, QMenu *menu, nm_menu_location_t loc, int at) {
+    NM_LOG("inject %d @ %d", loc, at);
+
+    NM_LOG("getting insertion point");
+
+    auto actions = menu->actions();
+    auto before = at < actions.count()
+        ? actions.at(at)
+        : nullptr;
+
+    if (before == nullptr)
+        NM_LOG("it seems the original item to add new ones before was never actually added to the menu (number of items when the action was created is %d, current is %d), appending to end instead", at, actions.count());
+
+    NM_LOG("checking for old items");
+
+    for (auto action : actions) {
+        if (action->property("nm_action") == true) {
+            return; // already added
+            /*menu->removeAction(action);
+            delete action;*/
+        }
+    }
+
+    NM_LOG("injecting new items");
+
+    // if it segfaults in createMenuTextItem, it's likely because
+    // AbstractNickelMenuController is invalid, which shouldn't happen while the
+    // menu which we added the signal from still can be shown... (but
+    // theoretically, it's possible)
 
     for (size_t i = 0; i < _items_n; i++) {
         nm_menu_item_t *it = _items[i];
-        if (it->loc == NM_MENU_LOCATION_MAIN_MENU && !ismm)
+        if (it->loc != loc)
             continue;
-        if (it->loc == NM_MENU_LOCATION_READER_MENU && !isrm)
+
+        NM_LOG("adding items '%s'...", it->lbl);
+
+        MenuTextItem* item = AbstractNickelMenuController_createMenuTextItem_orig(nmc, menu, QString::fromUtf8(it->lbl), false, false, "");
+        QAction* action = AbstractNickelMenuController_createAction_before(before, loc, i == _items_n-1, nmc, menu, item, true, true, true);
+
+        action->setProperty("nm_action", true);
+        QObject::connect(action, &QAction::triggered, [it](bool){ nm_menu_item_do(it); }); // note: we're capturing by value, i.e. the pointer to the global variable, rather then the stack variable, so this is safe
+    }
+}
+
+void nm_menu_item_do(nm_menu_item_t *it) {
```

Comment:
This function is identical to what was in the lambda which handles `QAction::triggered` for our custom actions.